### PR TITLE
feat(react-native-code-push): add react-native-code-push plugin

### PR DESCRIPTION
**Pull request by GSTJ**
# Why

This PR adds Codepush support without requiring manual changes to the native code. It's based on the work started on this thread, which never got merged: https://github.com/microsoft/react-native-code-push/pull/2415.

The community claimed this was broken on Android, so my work completely re-structured his work, ensured every step made sense, and followed the most up-to-date official Codepush documentation for both IOS and Android.

# How

This evolved from a fork of @deggertsen's work at [https://github.com/deggertsen/react-native-code-push-expo-plugin](https://github.com/deggertsen/react-native-code-push-expo-plugin/pulls). The work here never got published on NPM and had some outstanding issues.

This is a rewrite, where I followed the newest [Codepush Documentation](https://github.com/microsoft/react-native-code-push/blob/master/docs/) step-by-step and updated every file accordingly to match. I also refactored and consolidated his work to make it easy to understand and follow through.

Shoutout to @deggertsen for his solid base structure.

# Why not just merge it to Microsoft's repo?
@deggertsen has had a PR stuck on review since the beginning of the year, and there is a lot of bureaucracy in getting this code into CodePush's main repo.

Effectively, it isn't a repository made with this plugin in mind.

The idea here is to shine a light on the necessity of this plugin and battle-testing this approach with more users, and maybe we can get this merged into Codepush in the future.

@expo/config-plugins is community-maintained, so this fits better here for a head-start. I want to have this released on NPM sooner.

# Test Plan

I've created a react-native-code-push folder on /apps, so this plugin can be more easily tested.

1. Open the react-native-code-push folder on /apps
2. Run `yarn` to install the projects dependencies
3. Run `npx expo prebuild`
5. Check if the changes on the IOS and Android folders match the Codepush instructions here: [IOS](https://github.com/microsoft/react-native-code-push/blob/master/docs/setup-ios.md), [Android](https://github.com/microsoft/react-native-code-push/blob/master/docs/setup-android.md)

### CODE REVIEW

**Comment by deggertsen** [comment]
Love this! Thanks for building on this and getting it more out into the community.

**Comment by deggertsen** [comment]
We're using this on a Monorepo ourselves as a separate package. I think the main difference I'm seeing in the code in my monorepo is this line:
```
const codePushSettings = `include ':app', ':react-native-code-push'\nproject(':react-native-code-push').projectDir = new File(["node", "--print", "require.resolve('react-native-code-push/package.json')"].execute(null, rootDir).text.trim(), "../android/app")`
```

I feel like what you have here is a fantastic start. It will be easier to create PRs and suggest improvements once there is a baseline in place. So I don't feel like monorepo support is necessary out of the gate.

**Comment by GSTJ** [comment]
Thanks, @deggertsen, I'm with you there.

**Comment by GSTJ** [comment]
For anyone else facing this issue, I published this as an interim solution while this PR isn't merged: https://www.npmjs.com/package/react-native-code-push

**Comment by mvolonnino** [comment]
@GSTJ - wow this is awesome! My current app utilizes codepush right now, and we have been refactoring it to be able to leverage how great Expo has become - and with EAS updates being integrated and a great option, my company still wants to go with codepush for OTA stuff.

Is there plans for getting this merged / could I implement this in a production app?

**Comment by GSTJ** [comment]
@mvolonnino I'm currently using this version I published under NPM as an interim solution on our project; it's working great for us on production. The app has 300k+ active users, so that's been battle-tested.

> For anyone needing this plugin, I published this as an interim solution while this PR isn't merged: https://www.npmjs.com/package/react-native-code-push-plugin

**Comment by mvolonnino** [comment]
@GSTJ - wow thats amazing! Okay awesome, I am going to look into this this week and i'll let you guys know if I run into anything 👍

Thanks again for this, this is some amazing work and a big help to the rest of the community 🔥

**Comment by mvolonnino** [comment]
@GSTJ - quick question, i've followed the plugin readme, but just want to double check, after setting everything up through the plugin readme, I can now integrate CodePush through their docs correct?

Like wrapping the App, adding codepush options, etc?
https://learn.microsoft.com/en-us/appcenter/distribution/codepush/rn-plugin

**Comment by GSTJ** [comment]
@mvolonnino You can surely continue with their docs. The plugin only does the native changes for you.

About prebuild, I recommend you run it. It will apply the plugin changes.

You should be able to run the EAS build normally; yes, there is no need to change your CI.

**Comment by mvolonnino** [comment]
> About prebuild, I recommend you run it. It will apply the plugin changes.
> 
> You should be able to run the EAS build normally; yes, there is no need to change your CI.

@GSTJ - okay awesome! So just to make sure I understand you correctly, when i utilize the `eas build xxx` command to make a new build since i have app.config changes and native changes (through the plugin) - that is what you mean by you recommend running `expo prebuild` command to apply the plugin changes. The `eas build` command will generate a new build that I could utilize in the same regard

*edit - I see your edited post so I think that answered this follow up, but just double checking 🫡

**Comment by GSTJ** [comment]
@mvolonnino
**If you commit the 'IOS' and 'Android' folders to Git:**

- You need to run 'npx expo prebuild' every time you add a new plugin because adding a new plugin may change the contents of these folders.

**If you don't commit them:**

- The 'IOS' and 'Android' folders will be generated for you, as EAS Build runs `prebuild` on the CI automatically.
- Even though the CI handles this for you, it's still a good idea to run 'npx expo prebuild' locally to ensure that your local 'IOS' and 'Android' folders stay updated.

I recommend a read here https://docs.expo.dev/workflow/continuous-native-generation/

**Comment by mvolonnino** [comment]
@GSTJ - sweet thank you for the article, will read up on CNG as well. Thank you again. Appreciate all your help on this

**Comment by mgscreativa** [comment]
Hi @brentvatne can you take a brief look at this, this is really cool if it can get through. Thanks!

**Comment by brentvatne** [comment]
most of the expo team is out on holidays at the moment but we'll get back to you in the new year!

**Comment by mgscreativa** [comment]
Great! @brentvatne love you man, I really do! Happy holidays!

File: packages/react-native-code-push/src/android/mainApplicationDependency.ts
Comment (rvasseur31) [inline packages/react-native-code-push/src/android/mainApplicationDependency.ts:1]: We need to update this file as the MainApplication file is in kotlin now

Comment (GSTJ) [reply packages/react-native-code-push/src/android/mainApplicationDependency.ts:1]: I have already updated and supported Expo 50 at:
https://github.com/gstj/react-native-code-push-plugin
https://www.npmjs.com/package/react-native-code-push-plugin

As I'm not seeing much moviment on the Expo Team side to merge this, I'm not maintaining this PR.

If the Expo Teams wants this, just let me know and I'll promptly update the PR.

Comment (rvasseur31) [reply packages/react-native-code-push/src/android/mainApplicationDependency.ts:1]: Thanks for the link

**Comment by glncy** [comment]
Hopefully, Expo Team can see this and accept this. and hopefully expo team can add how to use codepush in expo router in their documentation. this is the solution i found on github using codepush and expo router: https://github.com/microsoft/react-native-code-push/pull/2415#issuecomment-1522585040

**Comment by brentvatne** [comment]
@glncy - codepush is being sunset by microsoft, so i don't think it makes sense for us to support this here. 

image

https://learn.microsoft.com/en-us/appcenter/distribution/codepush/

**Comment by LunatiqueCoder** [comment]
Hello @brentvatne 

CodePush will still work, but you will have to host it yourself on the server: 

https://github.com/microsoft/code-push-server

**Comment by brentvatne** [comment]
@LunatiqueCoder - it will kind of work yeah, but they do not intend to support the new architecture - so it's effectively dead in the water at this point since the new architecture is now default.

image
https://github.com/microsoft/react-native-code-push?tab=readme-ov-file#react-native-module-for-codepush

**Comment by LunatiqueCoder** [comment]
@brentvatne Oh that's cool. I missed that. Should be easy to convince the customers to migrate to EAS Updates now. Thanks!

**Pull request closed without merging**